### PR TITLE
Prioritize results from the current document if all other things are equal in a nav-to result

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Feedback from Platform team.  Right now we give no weighting to results in the current file or not when the results are identical (e.g. two methods with the same name in teh solution).

This is annoying for users who are in a particular file and want to quickly navigate to something in that file with an exact match name.

### CODE REVIEW

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.GeneratedDocumentSearch.cs [hunk 1]
```diff
@@ -56,7 +57,8 @@ internal abstract partial class AbstractNavigateToSearchService
 
             // First generate all the source-gen docs.  Then handoff to the standard search routine to find matches in them.  
             var generatedDocs = await project.GetSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false);
-            await ProcessDocumentsAsync(searchDocument: null, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, generatedDocs.ToSet<Document>(), cancellationToken).ConfigureAwait(false);
+            await ProcessDocumentsAsync(
+                searchDocument: null, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, generatedDocs.ToSet<Document>(), cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.GeneratedDocumentSearch.cs:61]:
```suggestion
            await ProcessDocumentsAsync(searchDocument: null, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, generatedDocs.ToSet<Document>(), cancellationToken).ConfigureAwait(false);
```

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs [hunk 2]
```diff
@@ -57,15 +57,17 @@ private static void ClearCachedData()
...
         {
             var solution = project.Solution;
-            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
-            var onItemFound = GetOnItemFoundCallback(solution, onResultFound, cancellationToken);
+            var onItemFound = GetOnItemFoundCallback(solution, activeDocument, onResultFound, cancellationToken);
 
             var documentKeys = project.Documents.SelectAsArray(DocumentKey.ToDocumentKey);
             var priorityDocumentKeys = priorityDocuments.SelectAsArray(DocumentKey.ToDocumentKey);
+
+            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.CachedDocumentSearch.cs:70]: reordered for consistency

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.GeneratedDocumentSearch.cs [hunk 3]
```diff
@@ -18,13 +18,14 @@ internal abstract partial class AbstractNavigateToSearchService
...
+            Document? activeDocument,
             Func<INavigateToSearchResult, Task> onResultFound,
             CancellationToken cancellationToken)
         {
             var solution = project.Solution;
-            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
-            var onItemFound = GetOnItemFoundCallback(solution, onResultFound, cancellationToken);
+            var onItemFound = GetOnItemFoundCallback(solution, activeDocument, onResultFound, cancellationToken);
 
+            var client = await RemoteHostClient.TryGetClientAsync(project, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.GeneratedDocumentSearch.cs:28]: reordered for consistency

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 4]
```diff
@@ -52,16 +52,23 @@ internal abstract partial class AbstractNavigateToSearchService
 
             // Prioritize the active documents if we have any.
             var highPriDocs = priorityDocuments.Where(d => project.ContainsDocument(d.Id)).ToSet();
-            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, highPriDocs, cancellationToken).ConfigureAwait(false);
+            await ProcessDocumentsAsync(
+                searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, highPriDocs, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:56]:
```suggestion
            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, highPriDocs, cancellationToken).ConfigureAwait(false);
```

File: src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs [hunk 5]
```diff
@@ -56,12 +56,18 @@ internal abstract partial class AbstractNavigateToSearchService
 
             // Then process non-priority documents.
             var lowPriDocs = project.Documents.Where(d => !highPriDocs.Contains(d)).ToSet();
-            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
+            await ProcessDocumentsAsync(
+                searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.InProcess.cs:60]:
```suggestion
            await ProcessDocumentsAsync(searchDocument, patternName, patternContainerOpt, declaredSymbolInfoKindsSet, onResultFound, lowPriDocs, cancellationToken).ConfigureAwait(false);
```

File: src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs [hunk 6]
```diff
@@ -95,14 +103,93 @@ private class NavigateToSearchResult : INavigateToSearchResult, INavigableItem
...
+            }
+
+            private string ComputeSecondarySort()
+            {
+                using var _ = ArrayBuilder<string>.GetInstance(out var parts);
+
+                // Ensure if all else is equal, that high-pri items (e.g. from the user's current file) come first
+                // before low pri items.  This only applies if things like the MatchKind are the same.  So we'll
+                // still show an exact match from another file before a substring match from the current file.
+                parts.Add(ComputeFolderDistance().ToString("X4"));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs:133]: this line (and the ComputeFolderDistance helper) is the meat of the change.

File: src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs [hunk 7]
```diff
@@ -189,25 +207,71 @@ private bool IsNamedType()
...
+
+            private string ComputeSecondarySort()
             {
-                get
+                using var _ = ArrayBuilder<string>.GetInstance(out var parts);
+
+                // Ensure if all else is equal, that high-pri items (e.g. from the user's current file) come first
+                // before low pri items.  This only applies if things like the MatchKind are the same.  So we'll
+                // still show an exact match from another file before a substring match from the current file.
+                parts.Add(ComputeFolderDistance().ToString("X4"));
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs:219]: this item (and the helper it calls) is the meat of hte change.

Comment (dibarbet) [reply src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs:219]: it seems like if we gave a file path with the result this is something the platform could do entirely on their end.  I think they would have to do this work for any search provider implemented via LSP as well (unless we add support for more complex sorting there too).

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/NavigateTo/RoslynNavigateToItem.cs:219]: Agreed. Though a virtue of this is that we can do this in a non allocating fashion. Def harder (but possible) on the platform side.

File: src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs [hunk 8]
```diff
@@ -1677,6 +1677,236 @@ public async Task FindRecordMember3(TestHost testHost, Composition composition)
...
+</Workspace>
+"), async workspace =>
+            {
+                _provider = CreateProvider(workspace);
+                _aggregator = new NavigateToTestAggregator(_provider);
+
+                var items = await _aggregator.GetItemsAsync("VisibleMethod");
+                var expectedItems = new List<NavigateToItem>()
+                {
+                    new NavigateToItem("VisibleMethod", NavigateToItemKind.Method, "csharp", null, null, s_emptyExactPatternMatch, null),
```
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/NavigateTo/NavigateToTests.cs:1718]: done.

**Comment by CyrusNajmabadi** [comment]
@dibarbet ptal.

**Review by dibarbet — APPROVED** [review]
Change seems good, what was the outcome of this?

> @dibarbet Discussing this with editor/platform now :)

**Comment by CyrusNajmabadi** [comment]
> Change seems good, what was the outcome of this?

Still continuing to discuss.  A little complex since we're also in the process of moving to all-in-one-search from go-to.  But it does seem reasonable.  We're going to bring this up with C++ as well.

**Pull request merged**